crates/ruff_python_formatter/src/verbatim.rs

```diff
@@ -573,33 +596,275 @@ impl Format<PyFormatContext<'_>> for TrailingFormatOffComment<'_> {
     }
 }
 
-struct VerbatimText(TextRange);
+/// Stores the indentation of a statement by storing the number of indentation characters.
+/// Storing the number of indentation characters is sufficient because:
+/// * Two indentations are equal if they result in the same column, regardless of the used tab size.
+///   This implementation makes use of this fact and assumes a tab size of 1.
+/// * The source document is correctly indented because it is valid Python code (or the formatter would have failed parsing the code).
+#[derive(Copy, Clone)]
+struct Indentation(u32);
+
+impl Indentation {
+    fn from_stmt(stmt: &Stmt, source: &str) -> Indentation {
+        let line_start = Locator::new(source).line_start(stmt.start());
+
+        let mut indentation = 0u32;
+        for c in source[TextRange::new(line_start, stmt.start())].chars() {
+            if matches!(c, ' ' | '\t') {
+                indentation += 1;
+            } else {
+                break;
+            }
+        }
+
+        Indentation(indentation)
+    }
+
+    fn trim_indent(self, ranged: impl Ranged, source: &str) -> TextRange {
+        let range = ranged.range();
+        let mut start_offset = TextSize::default();
+
+        for c in source[range].chars().take(self.0 as usize) {
+            if matches!(c, ' ' | '\t') {
+                start_offset += TextSize::new(1);
+            } else {
+                break;
+            }
+        }
+
+        TextRange::new(range.start() + start_offset, range.end())
+    }
+}
 
-fn verbatim_text<T>(item: T) -> VerbatimText
+/// Formats a verbatim range where the top-level nodes are statements (or statement-level comments).
+///
+/// Formats each statement as written in the source code, but adds the right indentation to match
+/// the indentation of formatted statements:
+///
+/// ```python
+/// def test():
+///   print("formatted")
+///   # fmt: off
+///   (
+///     not_formatted + b
+///   )
+///   # fmt: on
+/// ```
+///
+/// Gets formatted as
+///
+/// ```python
+/// def test():
+///     print("formatted")
+///     # fmt: off
+///     (
+///     not_formatted + b
+///     )
+///     # fmt: on
+/// ```
+///
+/// Notice how the `not_formatted + b` expression statement gets the same indentation as the `print` statement above,
+/// but the indentation of the expression remains unchanged. It changes the indentation to:
+/// * Prevent syntax errors because of different indentation levels between formatted and suppressed statements.
+/// * Align with the `fmt: skip` where statements are indented as well, but inner expressions are formatted as is.
```

Comment:
i like the docstring solution of only stripping the shared leading indentation, through docstring we would also already have the utils for that, except for own line comments.

Reply:
I would like that too. But we would need to track two different indentations:

* One to strip before statements
* The shared leading indentation, for comments

It may otherwise result in invalid code if a comment has less indentation than the statements. I think we consider introducing this complexity if we get reports that our fmt: off breaks the intended formatting (or e.g. test if the entire suite is disabled, and if so, don't fix up the indent)